### PR TITLE
Feat: Introduce the caching layer when loading SQL models

### PR DESCRIPTION
**Pull request by izeigerman**
There is also an opportunity to cache the `python_env` after it's been compiled by `prepare_env` but this hasn't been addressed in this PR.

### CODE REVIEW

**Comment by tobymao** [comment]
how do you invalidate the cache? since you're storing the ast dump, you probably need to invalidate it whenever the version changes

File: sqlmesh/core/loader.py [hunk 1]
```diff
@@ -165,18 +188,25 @@ def _load_sql_models(
...
-                with open(path, "r", encoding="utf-8") as file:
-                    try:
-                        expressions = parse(
-                            file.read(), default_dialect=config.model_defaults.dialect
-                        )
-                    except SqlglotError as ex:
-                        raise ConfigError(f"Failed to parse a model definition at '{path}': {ex}")
-                    model = load_model(
+
+                def _load() -> Model:
```
Comment (crericha) [inline sqlmesh/core/loader.py:199]: What is the cost of defining this method on every inner loop? From what I can tell, the method can be moved out of the inner loop if path is passed in as a param.

File: sqlmesh/core/loader.py [hunk 2]
```diff
@@ -186,11 +216,19 @@ def _load_sql_models(
...
                     )
-                    models[model.name] = model
 
-                    if isinstance(model, SeedModel):
-                        seed_path = model.seed_path
-                        self._track_file(seed_path)
+                cache_entry_name = "__".join(path.relative_to(models_path).parts).replace(
+                    ".sql", ""
+                )
+                cache_entry_id = self._model_cache_entry_id(path, context_path)
```
Comment (crericha) [inline sqlmesh/core/loader.py:223]: What do you think about encapsulating name and id into get_or_load, so the cache stuff is less exposed?

File: sqlmesh/dbt/loader.py [hunk 3]
```diff
@@ -55,19 +59,53 @@ def _load_models(
...
             context.jinja_macros.add_macros(
-                package.macros,
+                package.macro_infos,
                 package=package_name if package_name != context.project_name else None,
             )
+            macro_files_mtime = [
+                self._path_mtimes[m.path]
+                for m in package.macros.values()
+                if m.path in self._path_mtimes
+            ]
```
Comment (crericha) [inline sqlmesh/dbt/loader.py:73]: Why not move out of the loop and concat new times and do one max afterwards? Then you don't need the if/else for max.

File: sqlmesh/core/loader.py (hunk 1 above)
Comment (izeigerman) [reply sqlmesh/core/loader.py:199]: I don't think the cost matters in this case.

File: sqlmesh/core/loader.py (hunk 2 above)
Comment (izeigerman) [reply sqlmesh/core/loader.py:223]: Encapsulate how exactly? Please note that `name` is a key of the entry while `entry_id` is used to invalidate the entry

Comment (crericha) [reply sqlmesh/core/loader.py:223]: By passing in path and computing name and id  in get_or_load. model_cache init() would take in model_paths and context_path.

File: sqlmesh/core/loader.py [hunk 4]
```diff
@@ -79,6 +80,18 @@ def load(self, context: Context) -> LoadedProject:
         self._path_mtimes.clear()
         self._dag = DAG()
 
+        config_mtimes: t.Dict[Path, t.List[float]] = defaultdict(list)
+        for context_path, config in self._context.configs.items():
+            for config_file in context_path.glob("config.*"):
+                self._track_file(config_file)
+                config_mtimes[context_path].append(self._path_mtimes[config_file])
```
Comment (crericha) [inline sqlmesh/core/loader.py:87]: If you want, you could augment _track_file to take a type like self._track_file(config_file, FileType.Config) and then add helpers get_mtimes(FileType.Config). Not sure its worth it, but I'll continue to throw out ideas to reduce caching specific code from the loader and you can do what you want with them. :-)

File: sqlmesh/core/model/cache.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+from sqlmesh.utils.cache import FileCache
+from sqlmesh.utils.pydantic import PydanticModel
+
+
+class SqlModelCacheEntry(PydanticModel):
+    model: SqlModel
+    rendered_query: t.Dict
+
+
+class ModelCache:
```
Comment (tobymao) [inline sqlmesh/core/model/cache.py:18]: let's add a few doc strings

File: sqlmesh/dbt/loader.py [hunk 6]
```diff
@@ -55,19 +59,38 @@ def _load_models(
...
             )
+            macros_mtimes.extend(
+                [
+                    self._path_mtimes[m.path]
+                    for m in package.macros.values()
+                    if m.path in self._path_mtimes
+                ]
+            )
+
+        macros_max_mtime: t.Optional[float] = max(macros_mtimes) if macros_mtimes else None
```
Comment (tobymao) [inline sqlmesh/dbt/loader.py:77]: why do you need the type here

File: sqlmesh/dbt/loader.py [hunk 7]
```diff
@@ -77,11 +100,110 @@ def _load_models(
...
+                yaml_mtime = self._path_mtimes.get(nested)
+                if yaml_mtime:
+                    max_mtime = max(max_mtime, yaml_mtime) if max_mtime is not None else yaml_mtime
+
+        if max_mtime is not None:
+            result[root] = max_mtime
+
+        return result
+
+    class _Cache:
```
Comment (tobymao) [inline sqlmesh/dbt/loader.py:138]: damn dbt has it's only cache ?

File: sqlmesh/utils/cache.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        return None
+
+    def put(self, name: str, entry_id: str, value: T) -> None:
+        if not self._path.exists():
+            self._path.mkdir(parents=True)
+        if not self._path.is_dir():
+            raise SQLMeshError(f"Cache path '{self._path}' is not a directory.")
+
+        for obsolete_cache_file in self._path.glob(f"{name}__*"):
+            obsolete_cache_file.unlink()
```
Comment (tobymao) [inline sqlmesh/utils/cache.py:60]: this deletes all old files?

**Review by tobymao — APPROVED** [review]

File: sqlmesh/utils/cache.py (hunk 8 above)
Comment (izeigerman) [reply sqlmesh/utils/cache.py:60]: that's correct

File: sqlmesh/dbt/loader.py (hunk 7 above)
Comment (izeigerman) [reply sqlmesh/dbt/loader.py:138]: sorry, not sure I understand the comment

Comment (tobymao) [reply sqlmesh/dbt/loader.py:138]: just complaining that it seems like we needed two similar classes

Comment (izeigerman) [reply sqlmesh/dbt/loader.py:138]: yeah, there's potentially a way to consolidate them but doesn't seem like it's worth the investment right now.

File: sqlmesh/core/loader.py [hunk 9]
```diff
@@ -265,3 +296,31 @@ def _glob_paths(
...
+            )
+
+        def _model_cache_entry_id(self, model_path: Path) -> str:
+            mtimes = [
+                self._loader._path_mtimes[model_path],
+                self._loader._macros_max_mtime,
+                self._loader._config_mtimes.get(self._context_path),
+                self._loader._config_mtimes.get(self._loader._context.sqlmesh_path),
+            ]
+            return str(int(max([m for m in mtimes if m is not None])))
```
Comment (vchan) [inline sqlmesh/core/loader.py:326]:
```suggestion
            return str(int(max(m for m in mtimes if m is not None)))
```

Comment (izeigerman) [reply sqlmesh/core/loader.py]: good catch, thanks!

**Pull request merged**